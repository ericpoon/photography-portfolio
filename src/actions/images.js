import firebase from '../firebase/firebase';

const database = firebase.database();
const storage = firebase.storage();

export const setImages = images => ({
  type: 'SET_IMAGES',
  images,
});

export const startDownloadImages = () => {
  return (dispatch) => {
    return database.ref('/images').once('value')
      .then((snapshot) => {
        const images = [];
        snapshot.forEach((childSnapshot) => {
          images.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setImages(images));
      });
  };
};

export const addImage = image => ({
  type: 'ADD_IMAGE',
  image,
});

export const startAddImage = (image) => {
  const {
    title, subtitle, description, file,
  } = image;
  return (dispatch) => {
    const id = database.ref('/images').push().key;
    return storage.ref(`/images/${id}`).put(file, { cacheControl: 'private, max-age=60' })
      .then((snapshot) => {
        const fileUrl = snapshot.downloadURL;
        return database.ref(`/images/${id}`).set({
          title, subtitle, description, fileUrl,
        })
          .then(() => {
            dispatch(addImage({
              id,
              title,
              subtitle,
              description,
              fileUrl,
            }));
          });
      });
  };
};

export const editImage = (id, updates) => ({
  type: 'EDIT_IMAGE',
  id,
  updates,
});

export const startEditImage = (id, updates) => {
  return (dispatch) => {
    return database.ref(`/images/${id}`).update(updates)
      .then(() => {
        dispatch(editImage(id, updates));
      });
  };
};

export const deleteImage = id => ({
  type: 'DELETE_IMAGE',
  id,
});

export const startDeleteImage = (id) => {
  return (dispatch) => {
    return database.ref(`images/${id}`).remove()
      .then(() => {
        dispatch(deleteImage(id));
      });
  };
};
