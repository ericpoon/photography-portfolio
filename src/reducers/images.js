const initialState = [];

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_IMAGES':
      return action.images;
    case 'ADD_IMAGE':
      return [...state, action.image];
    case 'EDIT_IMAGE':
      return state.map((i) => {
        return i.id === action.id ? { ...i, ...action.updates } : i;
      });
    case 'DELETE_IMAGE':
      return state.filter(i => i.id !== action.id);
    default:
      return state;
  }
};
