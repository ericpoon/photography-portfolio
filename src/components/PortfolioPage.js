import React from 'react';
import { connect } from 'react-redux';
import { showDetail } from '../actions/gallery';
import ImageDetailModal from './ImageDetailModal';
import ImageCardCollection from './ImageCardCollection';
import ImageCard from './ImageCard';

const PortfolioPage = props => (
  <div>
    <ImageDetailModal />
    <div className={'content-container'}>
      <ImageCardCollection>
        {props.images.map((image) => {
          return <ImageCard key={image.id} {...image} onImageClick={() => props.showDetail()} />;
        })}
      </ImageCardCollection>
    </div>
  </div>
);

const mapStateToProps = state => ({
  images: state.images,
});

export default connect(mapStateToProps, { showDetail })(PortfolioPage);
