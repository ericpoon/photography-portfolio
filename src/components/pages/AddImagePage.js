import React from 'react';
import { connect } from 'react-redux';
import { startAddImage } from '../../actions/images';
import ImageCard from '../ImageCard';

export const AddImagePage = (props) => {
  return (
    <div className={'content-container content-container--small'}>
      <ImageCard
        editable
        onSaveClick={updates => props.startAddImage(updates)}
      />
    </div>
  );
};

export default connect(undefined, { startAddImage })(AddImagePage);
