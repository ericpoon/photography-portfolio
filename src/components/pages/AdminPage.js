import React from 'react';
import { connect } from 'react-redux';
import { startEditImage, startDeleteImage } from '../../actions/images';
import ImageCardCollection from '../ImageCardCollection';
import ImageCard from '../ImageCard';

export const AdminPage = (props) => {
  return (
    <div>
      <div className={'content-container'}>
        <ImageCardCollection>
          {props.images.map((image) => {
            return (
              <ImageCard
                {...image}
                editable
                key={image.id}
                onSaveClick={updates => props.startEditImage(image.id, updates)}
                onDeleteClick={() => props.startDeleteImage(image.id)}
              />);
          })}
        </ImageCardCollection>
      </div>
    </div>
  );
};

const mapStateToProps = state => ({
  images: state.images,
});

export default connect(mapStateToProps, { startEditImage, startDeleteImage })(AdminPage);
