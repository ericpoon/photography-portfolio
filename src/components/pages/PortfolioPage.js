import React from 'react';
import { connect } from 'react-redux';
import { showDetail } from '../../actions/gallery';
import ImageDetailModal from '../ImageDetailModal';
import ImageBoard from '../ImageBoard';
import ImageCard from '../presentational/ImageCard';

export const PortfolioPage = props => (
  <div>
    <ImageDetailModal />
    <div className={'content-container'}>
      <ImageBoard>
        {props.images.map((image) => {
          return (
            <ImageCard
              {...image}
              key={image.id}
              onImageClick={() => props.showDetail(image.id)}
            />
          );
        })}
      </ImageBoard>
    </div>
  </div>
);

const mapStateToProps = state => ({
  images: state.images,
});

export default connect(mapStateToProps, { showDetail })(PortfolioPage);
