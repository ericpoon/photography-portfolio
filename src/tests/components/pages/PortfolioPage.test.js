import React from 'react';
import { shallow } from 'enzyme';
import { PortfolioPage } from '../../../components/pages/PortfolioPage';
import images from '../../fixtures/images';

let wrapper,
  showDetail;

beforeEach(() => {
  showDetail = jest.fn();
  wrapper = shallow(<PortfolioPage images={images} showDetail={showDetail} />);
});

it('should render PortfolioPage', () => {
  expect(wrapper).toMatchSnapshot();
});

it('should have correct number of image cards', () => {
  expect(wrapper.find('Connect(ImageCard)').length).toBe(images.length);
});

it('should have no editable image cards', () => {
  expect(wrapper
    .find('Connect(ImageCard)')
    .filterWhere(imageCard => imageCard.prop('editable'))
    .length)
    .toBe(0);
});

it('should show image detail after clicking image', () => {
  const firstImageCard = wrapper.find('Connect(ImageCard)').at(0);
  firstImageCard.prop('onImageClick')();
  expect(showDetail).toHaveBeenCalledWith(firstImageCard.prop('id'));
});
