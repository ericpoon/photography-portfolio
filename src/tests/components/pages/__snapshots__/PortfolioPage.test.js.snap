// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PortfolioPage 1`] = `
<div>
  <Connect(ImageDetailModal) />
  <div
    className="content-container"
  >
    <Component>
      <ImageCard
        description="Putarem quodque tamquam ii ob an deumque. Me assequi sapores ab verarum. Mei qui religionis nam fas devenietur ponderibus. Junctae invenit plausum differt de ipsamet externo is. Mutationum sui negationem exponuntur"
        fileUrl="www.google.com"
        id={1000}
        key="1000"
        onImageClick={[Function]}
        subtitle="Dec 31, 2017"
        title="Kowloon"
      />
      <ImageCard
        description="Veniebant sex videbatur assignare eas. Ima secundum synopsis imponere spectant. realitatem aliquamdiu hoc eam requiratur pro."
        fileUrl="www.google.com"
        id={1001}
        key="1001"
        onImageClick={[Function]}
        subtitle="Jan 18, 2018"
        title="Hong Kong Island"
      />
      <ImageCard
        description="Oculis nondum fusius sub ens urgeat fuerit rea sum. Paulo his novas somno audio nia. Religio lor ordinis propria habemus cui. Cohaereant vis sim consistere ope concipitur imaginabor afferuntur."
        fileUrl="www.google.com"
        id={1002}
        key="1002"
        onImageClick={[Function]}
        subtitle="Jan 1, 2018"
        title="Causeway Bay"
      />
      <ImageCard
        description="Eoex deumque reddere publice ea similes attinet. Sensuum acgifallant incipit."
        fileUrl="www.google.com"
        id={1003}
        key="1003"
        onImageClick={[Function]}
        subtitle="April 27, 2017"
        title="Shenzhen"
      />
      <ImageCard
        description="Vocem id alias to curam multi brevi in novam ii. Ima secundum synopsis."
        fileUrl="www.google.com"
        id={1004}
        key="1004"
        onImageClick={[Function]}
        subtitle="Oct 9, 2016"
        title="Shanghai"
      />
    </Component>
  </div>
</div>
`;
